xcm/pallet-xcm/src/lib.rs

```diff
@@ -579,8 +922,71 @@ pub mod pallet {
 				None
 			}
 		}
+
+		/// Note that a particular destination to whom we would like to send a message is unknown
+		/// and queue it for version discovery.
+		fn note_unknown_version(dest: &MultiLocation) {
+			let versioned_dest = VersionedMultiLocation::from(dest.clone());
+			VersionDiscoveryQueue::<T>::mutate(|q| {
+				if let Some(index) = q.iter().position(|i| &i.0 == &versioned_dest) {
+					// exists - just bump the count.
+					q[index].1.saturating_inc();
+				} else {
+					let _ = q.try_push((versioned_dest, 1));
+				}
+			});
+		}
+	}
+
+	impl<T: Config> WrapVersion for Pallet<T> {
+		fn wrap_version<Call>(
+			dest: &MultiLocation,
+			xcm: impl Into<VersionedXcm<Call>>,
+		) -> Result<VersionedXcm<Call>, ()> {
+			SupportedVersion::<T>::get(XCM_VERSION, LatestVersionedMultiLocation(dest))
+				.or_else(|| {
+					Self::note_unknown_version(dest);
+					SafeXcmVersion::<T>::get()
+				})
+				.ok_or(())
+				.and_then(|v| xcm.into().into_version(v.min(XCM_VERSION)))
+		}
 	}
 
+	impl<T: Config> VersionChangeNotifier for Pallet<T> {
+		/// Start notifying `location` should the XCM version of this chain change.
+		///
+		/// When it does, this type should ensure an `QueryResponse` message is sent with the given
+		/// `query_id` & `max_weight` and with a `response` of `Repsonse::Version`. This should happen
+		/// until/unless `stop` is called with the correct `query_id`.
+		///
+		/// If the `location` has an ongoing notification and when this function is called, then an
+		/// error should be returned.
+		fn start(dest: &MultiLocation, query_id: QueryId, max_weight: u64) -> XcmResult {
+			let versioned_dest = LatestVersionedMultiLocation(dest);
+			let already = VersionNotifyTargets::<T>::contains_key(XCM_VERSION, versioned_dest);
+			ensure!(!already, XcmError::InvalidLocation);
+
+			let xcm_version = T::AdvertizeXcmVersion::get();
+			let response = Response::Version(xcm_version);
+			let instruction = QueryResponse { query_id, response, max_weight };
+			T::XcmRouter::send_xcm(dest.clone(), Xcm(vec![instruction]))?;
+
+			let value = (query_id, max_weight, xcm_version);
+			VersionNotifyTargets::<T>::insert(XCM_VERSION, versioned_dest, value);
+			Ok(())
+		}
+
+		/// Stop notifying `location` should the XCM change. Returns an error if there is no existing
+		/// notification set up.
+		fn stop(dest: &MultiLocation) -> XcmResult {
+			let versioned_dest = LatestVersionedMultiLocation(dest);
+			let already = VersionNotifyTargets::<T>::contains_key(XCM_VERSION, versioned_dest);
+			ensure!(already, XcmError::InvalidLocation);
+			VersionNotifyTargets::<T>::remove(XCM_VERSION, versioned_dest);
```

Comment:
im not convinced this is preferable to just removing the value. the only upside is that there wont be the logical path of a DB write in the case of no value existing, but that will actually not be that expensive since it will not modify the merkle trie when there is nothing to update.